### PR TITLE
feat: dWallet network key version

### PR DESCRIPTION
**Pull request by yaelAbergel**

### CODE REVIEW

File: crates/pera-core/src/authority/authority_per_epoch_store.rs [hunk 1]
```diff
@@ -1047,43 +1049,47 @@ impl AuthorityPerEpochStore {
...
+    pub fn get_encryption_of_decryption_key_shares(
+        &self,
+    ) -> DwalletMPCResult<HashMap<u8, Vec<EncryptionOfNetworkDecryptionKeyShares>>> {
+        // Todo (#396): Read the decryption key share from the network DKG output
         if self.epoch() == FIRST_EPOCH_ID {
-            return Err(PeraError::Unknown(
-                "First epoch does not have decryption key shares, need to run network DKG"
-                    .to_string(),
-            ));
+            return Err(DwalletMPCError::WrongEpoch(self.epoch()));
```
Comment (ItayLevyOfficial) [inline crates/pera-core/src/authority/authority_per_epoch_store.rs:1057]: This is not true, as there can be a scenario where the chain's secret has already been generated also in the first epoch. Please add a todo here.

Comment (ItayLevyOfficial) [inline crates/pera-core/src/authority/authority_per_epoch_store.rs:1057]: This is not true, as there can be a scenario where the chain's secret has already been generated also in the first epoch. Please add a todo here.

File: crates/pera-core/src/authority/authority_per_epoch_store.rs [hunk 2]
```diff
@@ -1047,43 +1049,47 @@ impl AuthorityPerEpochStore {
...
+        &self,
+        key_type: DWalletMPCNetworkKey,
+    ) -> DwalletMPCResult<Vec<u8>> {
+        // Todo (#396): Read the decryption key share from the network DKG output
         if self.epoch() == FIRST_EPOCH_ID {
-            return Err(PeraError::Unknown(
-                "First epoch does not have decryption key shares, need to run network DKG"
-                    .to_string(),
-            ));
+            return Err(return Err(DwalletMPCError::WrongEpoch(self.epoch())));
```
Comment (ItayLevyOfficial) [inline crates/pera-core/src/authority/authority_per_epoch_store.rs:1079]: same as above

File: crates/pera-core/src/dwallet_mpc/mod.rs [hunk 3]
```diff
@@ -343,6 +351,7 @@ pub(crate) fn from_event(
     event: &Event,
     dwallet_mpc_manager: &DWalletMPCManager,
     party_id: PartyID,
+    dwallet_network_key_version: u8,
```
Comment (ItayLevyOfficial) [inline crates/pera-core/src/dwallet_mpc/mod.rs:354]: you should have this field in the manager's state, and not pass it as a parameter to this function

File: crates/pera-core/src/dwallet_mpc/mpc_manager.rs [hunk 4]
```diff
@@ -64,7 +64,7 @@ pub struct DWalletMPCManager {
     pub(crate) malicious_actors: HashSet<AuthorityName>,
     weighted_threshold_access_structure: WeightedThresholdAccessStructure,
     weighted_parties: HashMap<PartyID, PartyID>,
-    outputs_manager: DWalletMPCOutputsVerifier,
+    pub outputs_manager: DWalletMPCOutputsVerifier,
```
Comment (ItayLevyOfficial) [inline crates/pera-core/src/dwallet_mpc/mpc_manager.rs:67]: why should it be public? I wanted it to be internal only to detect malicious actors

File: crates/pera-core/src/dwallet_mpc/mpc_outputs_verifier.rs [hunk 5]
```diff
@@ -61,9 +64,23 @@ impl DWalletMPCOutputsVerifier {
...
+            {
+                versions.len() as u8
+            } else {
+                1
+            },
         }
     }
 
+    pub fn network_key_version(&self) -> u8 {
+        self.network_key_version
```
Comment (ItayLevyOfficial) [inline crates/pera-core/src/dwallet_mpc/mpc_outputs_verifier.rs:81]: Why are you not simply making this field public?

File: sdk/typescript/src/dwallet-mpc/dkg.ts [hunk 6]
```diff
@@ -40,13 +40,15 @@ export interface DWallet {
 	session_id: string;
 	dwallet_cap_id: string;
 	output: number[];
+	dwallet_mpc_network_key_version: number;
```
Comment (ItayLevyOfficial) [inline sdk/typescript/src/dwallet-mpc/dkg.ts:43]:
```suggestion
	network_key_version: number;
```

File: sdk/typescript/src/dwallet-mpc/dkg.ts [hunk 7]
```diff
@@ -40,13 +40,15 @@ export interface DWallet {
...
 }
 
 export interface CreatedDwallet {
 	id: string;
 	centralizedDKGOutput: number[];
 	decentralizedDKGOutput: number[];
 	dwalletCapID: string;
+	dwalletMPCNetworkKeyVersion: number;
```
Comment (ItayLevyOfficial) [inline sdk/typescript/src/dwallet-mpc/dkg.ts:51]:
```suggestion
	networkKeyVersion: number;
```

File: crates/pera-types/src/pera_system_state/pera_system_state_inner_v2.rs [hunk 8]
```diff
@@ -59,7 +60,7 @@ pub struct PeraSystemStateInnerV2 {
     pub epoch: u64,
     pub protocol_version: u64,
     pub system_state_version: u64,
-    pub encrypted_decryption_key_share: Vec<Vec<u8>>,
+    pub encrypted_decryption_key_share: VecMap<u8, Vec<EncryptionOfNetworkDecryptionKeyShares>>,
```
Comment (ItayLevyOfficial) [inline crates/pera-types/src/pera_system_state/pera_system_state_inner_v2.rs:63]: Q: Why using a `VecMap` here is a good idea?

File: crates/pera-framework/packages/pera-system/sources/dwallet_network_key.move [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    }
+
+    /// Struct to store the network encryption of decryption key shares
+    public struct EncryptionOfNetworkDecryptionKeyShares has store, copy {
+        epoch: u64,
+        current_epoch_shares: vector<vector<u8>>,
+        previous_epoch_shares: vector<vector<u8>>,
+    }
+
+    public fun new_encrypted_network_decryption_key_shares(epoch: u64, current_epoch_shares: vector<vector<u8>>, previous_epoch_shares: vector<vector<u8>>): EncryptionOfNetworkDecryptionKeyShares {
```
Comment (ItayLevyOfficial) [inline crates/pera-framework/packages/pera-system/sources/dwallet_network_key.move:20]: Docs docs docs. Wtf those function do?

File: crates/pera-framework/packages/pera-system/sources/pera_system_state_inner.move [hunk 10]
```diff
@@ -153,7 +159,7 @@ module pera_system::pera_system_state_inner {
         system_state_version: u64,
         /// These are the encrypted decryption key shares for the current epoch, used for dWallet MPC session.
         /// The shares are indexed by the validator index of the current epoch committee.
-        encrypted_decryption_key_shares: vector<vector<u8>>,
+        encryption_of_decryption_key_shares: VecMap<u8, vector<EncryptionOfNetworkDecryptionKeyShares>>,
```
Comment (ItayLevyOfficial) [inline crates/pera-framework/packages/pera-system/sources/pera_system_state_inner.move:162]: update docs

File: crates/pera-framework/packages/pera-system/sources/pera_system_state_inner.move [hunk 11]
```diff
@@ -351,12 +357,39 @@ module pera_system::pera_system_state_inner {
...
+
+        if (self.encryption_of_decryption_key_shares.contains(&key_type)) {
+            self.encryption_of_decryption_key_shares.get_mut(&key_type).push_back(new_version);
+            return
+        };
+
+        self.encryption_of_decryption_key_shares.insert(key_type, vector[new_version]);
     }
 
+    public(package) fun store_encryption_of_decryption_key_shares(
```
Comment (ItayLevyOfficial) [inline crates/pera-framework/packages/pera-system/sources/pera_system_state_inner.move:372]: docs

File: crates/pera-framework/packages/pera-system/sources/pera_system_state_inner.move [hunk 12]
```diff
@@ -351,12 +357,39 @@ module pera_system::pera_system_state_inner {
         self.validators.lock_next_epoch_committee(self.epoch);
     }
 
-    public(package) fun store_encrypted_decryption_key_shares(self: &mut PeraSystemStateInnerV2, shares: vector<vector<u8>>) {
-        self.encrypted_decryption_key_shares = shares;
+    public(package) fun new_encryption_of_decryption_key_shares_version(self: &mut PeraSystemStateInnerV2, shares: vector<vector<u8>>, key_type: u8) {
```
Comment (ItayLevyOfficial) [inline crates/pera-framework/packages/pera-system/sources/pera_system_state_inner.move:360]: docs

File: crates/pera-types/src/messages_dwallet_mpc.rs [hunk 13]
```diff
@@ -11,7 +12,7 @@ pub enum MPCRound {
     /// The first round of the DKG protocol.
     DKGFirst,
     /// The second round of the DKG protocol.
-    DKGSecond,
+    DKGSecond(u8),
```
Comment (ItayLevyOfficial) [inline crates/pera-types/src/messages_dwallet_mpc.rs:15]: maybe doc this weird u8

**Review by zeevmoney — APPROVED** [review]

File: crates/pera-core/src/authority/authority_per_epoch_store.rs (hunk 1 above)
Comment (yaelAbergel) [reply crates/pera-core/src/authority/authority_per_epoch_store.rs:1059]: Fixed in later PR

File: crates/pera-core/src/dwallet_mpc/mpc_outputs_verifier.rs (hunk 5 above)
Comment (yaelAbergel) [reply crates/pera-core/src/dwallet_mpc/mpc_outputs_verifier.rs:81]: This is treated differently in a later PR

File: sdk/typescript/src/dwallet-mpc/dkg.ts (hunk 6 above)
Comment (yaelAbergel) [reply sdk/typescript/src/dwallet-mpc/dkg.ts:43]: There is another key named as "network key" within the sui code

File: crates/pera-framework/packages/pera-system/sources/pera_system_state_inner.move (hunk 10 above)
Comment (yaelAbergel) [reply crates/pera-framework/packages/pera-system/sources/pera_system_state_inner.move:162]: fixed in later PR

File: crates/pera-framework/packages/pera-system/sources/pera_system_state_inner.move (hunk 11 above)
Comment (yaelAbergel) [reply crates/pera-framework/packages/pera-system/sources/pera_system_state_inner.move:372]: added in later PR

File: crates/pera-types/src/messages_dwallet_mpc.rs (hunk 13 above)
Comment (yaelAbergel) [reply crates/pera-types/src/messages_dwallet_mpc.rs:15]: I use alias in later PR

**Pull request merged**